### PR TITLE
add Subgraph executor & refactor subgraph

### PR DESCRIPTION
**Pull request by nevermore3**
The expand to subgraph procedure expands to subgraph nodes reachable from the start node following relationships to max-level by user defined  Returns the collection of nodes in the subgraph, and the collection of relationships by user defined between all subgraph nodes

### CODE REVIEW

File: src/executor/query/DataCollectExecutor.cpp [hunk 1]
```diff
@@ -61,12 +61,15 @@ Status DataCollectExecutor::collectSubgraph(const std::vector<std::string>& vars
     DataSet ds;
     ds.colNames = std::move(colNames_);
     // the subgraph not need duplicate vertices or edges, so dedup here directly
-    std::unordered_set<Value> uniqueVids;
+    std::unordered_set<std::string> uniqueVids;
```
Comment (Shylock-Hg) [inline src/executor/query/DataCollectExecutor.cpp:64]: Why change this?

Comment (nevermore3) [reply src/executor/query/DataCollectExecutor.cpp:64]: According to the benchmark of yee， The efficiency of hash(string) is higher than that of hash(value)

Comment (Shylock-Hg) [reply src/executor/query/DataCollectExecutor.cpp:64]: Yes, but the Value(int).toString() is not efficient, I think you could use the int binary directly.

Comment (nevermore3) [reply src/executor/query/DataCollectExecutor.cpp:64]: Do you mean  Value.toInt()

Comment (Shylock-Hg) [reply src/executor/query/DataCollectExecutor.cpp:64]: Yes.

Comment (nevermore3) [reply src/executor/query/DataCollectExecutor.cpp:64]: Value.toInt() will get wrong result when Value is string 
eg: Value("123 abc").toInt == Value("123 cde").toInt

Comment (Shylock-Hg) [reply src/executor/query/DataCollectExecutor.cpp:64]: toString when value is string

File: src/planner/Algo.h [hunk 2]
```diff
@@ -139,6 +139,45 @@ class CartesianProduct final : public SingleDependencyNode {
...
+
+    uint32_t steps() const {
+        return steps_;
+    }
+
+private:
+    Subgraph(QueryContext* qctx,
+             PlanNode* input,
+             std::string oneMoreStepOutput,
+             std::string currentStepVar,
```
Comment (yixinglu) [inline src/planner/Algo.h:169]: const std::string &

File: src/executor/algo/SubgraphExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    return finish(ResultBuilder().value(Value(std::move(ds))).finish());
+}
+
+void SubgraphExecutor::oneMoreStep() {
+    auto* subgraph = asNode<Subgraph>(node());
+    auto output = subgraph->oneMoreStepOutput();
+    VLOG(1) << "OneMoreStep Input: " << subgraph->inputVar() << " Output: " << output;
+    auto iter = ectx_->getResult(subgraph->inputVar()).iter();
+    DCHECK(iter->isGetNeighborsIter());
+    DCHECK(!!iter);
```
Comment (yixinglu) [inline src/executor/algo/SubgraphExecutor.cpp:62]: ditto

File: src/validator/TraversalValidator.cpp [hunk 4]
```diff
@@ -199,13 +199,14 @@ PlanNode* TraversalValidator::buildRuntimeInput(Starts& starts, PlanNode*& proje
 Expression* TraversalValidator::buildNStepLoopCondition(uint32_t steps) const {
     VLOG(1) << "steps: " << steps;
     // ++loopSteps{0} <= steps
-    auto loopSteps = vctx_->anonVarGen()->getVar();
-    qctx_->ectx()->setValue(loopSteps, 0);
+    if (LIKELY(loopSteps_.empty())) {
+        loopSteps_ = vctx_->anonVarGen()->getVar();
+    }
```
Comment (yixinglu) [inline src/validator/TraversalValidator.cpp:204]: Why not initialize this variable in constructor?

File: src/validator/TraversalValidator.h [hunk 5]
```diff
@@ -68,6 +68,7 @@ class TraversalValidator : public Validator {
     Steps                 steps_;
     std::string           srcVidColName_;
     PlanNode*             projectStartVid_{nullptr};
+    mutable std::string   loopSteps_;
```
Comment (yixinglu) [inline src/validator/TraversalValidator.h:71]: as i said above, here need not to make this variable mutable.

File: src/executor/query/DataCollectExecutor.cpp [hunk 6]
```diff
@@ -62,11 +62,14 @@ Status DataCollectExecutor::collectSubgraph(const std::vector<std::string>& vars
     ds.colNames = std::move(colNames_);
     // the subgraph not need duplicate vertices or edges, so dedup here directly
     std::unordered_set<Value> uniqueVids;
-    std::unordered_set<std::tuple<Value, std::string, int64_t, Value>> uniqueEdges;
-    for (auto& var : vars) {
-        auto& hist = ectx_->getHistory(var);
-        for (auto& result : hist) {
-            auto iter = result.iter();
+    std::unordered_set<std::tuple<Value, EdgeType, EdgeRanking, Value>> uniqueEdges;
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply src/executor/query/DataCollectExecutor.cpp:65]: hash<tuple> is more faster than hash(Edge)

File: src/executor/query/DataCollectExecutor.cpp [hunk 7]
```diff
@@ -62,11 +62,14 @@ Status DataCollectExecutor::collectSubgraph(const std::vector<std::string>& vars
...
+    for (size_t i = 0; i < vars.size(); ++i) {
```
Comment (monadbobo) [inline src/executor/query/DataCollectExecutor.cpp:66]: It would be better to use an iterator.

File: src/executor/query/DataCollectExecutor.cpp [hunk 8]
```diff
@@ -62,11 +62,14 @@ Status DataCollectExecutor::collectSubgraph(const std::vector<std::string>& vars
...
+        const auto& hist = ectx_->getHistory(vars[i]);
+        for (size_t j = 0; j < hist.size(); ++j) {
+            if (i == 0 && j == hist.size() - 1) {
```
Comment (monadbobo) [inline src/executor/query/DataCollectExecutor.cpp:69]: ditto

File: src/executor/algo/SubgraphExecutor.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    if (currentStep == steps) {
+        oneMoreStep();
+        return finish(ResultBuilder().value(Value(std::move(ds))).finish());
+    }
+
+    VLOG(1) << "input: " << subgraph->inputVar() << " output: " << node()->outputVar();
+    auto iter = ectx_->getResult(subgraph->inputVar()).iter();
+    DCHECK(iter->isGetNeighborsIter());
+    DCHECK(!!iter);
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply src/executor/algo/SubgraphExecutor.cpp:34]: done

File: src/planner/Algo.h (hunk 2 above)
Comment (nevermore3) [reply src/planner/Algo.h:169]: done

File: src/validator/TraversalValidator.cpp (hunk 4 above)
Comment (nevermore3) [reply src/validator/TraversalValidator.cpp:204]: done

File: src/executor/query/DataCollectExecutor.cpp (hunk 7 above)
Comment (nevermore3) [reply src/executor/query/DataCollectExecutor.cpp:66]: ok

**Review by yixinglu — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**